File: xml-fluss-apt/src/main/java/xmlfluss/apt/XmlDslProcessor.java
package xmlfluss.apt;

import xmlfluss.apt.spi.AptSymbolProvider;
import xmlfluss.codegen.classify.CoreClassifier;
import xmlfluss.codegen.model.NestedRegistry;
import xmlfluss.codegen.model.RecordSpec;
import xmlfluss.codegen.plan.DispatchPlan;
import xmlfluss.codegen.plan.DispatchPlanBuilder;
import xmlfluss.codegen.spi.RecordSymbol;

import javax.annotation.processing.AbstractProcessor;
import javax.annotation.processing.RoundEnvironment;
import javax.annotation.processing.SupportedAnnotationTypes;
import javax.annotation.processing.SupportedSourceVersion;
import javax.lang.model.SourceVersion;
import javax.lang.model.element.TypeElement;
import javax.tools.Diagnostic;
import java.util.Set;

/**
 * APT processor that generates a streaming parser for every record annotated with
 * {@code @xmlfluss.XmlRecord}. The generated parser exposes
 * {@code public static java.util.stream.Stream<T> parse(InputStream)} and
 * {@code public static Stream<T> parse(InputStream, boolean ignoreNamespace)}.
 *
 * <p>The MVP supports {@code @XmlRecord}, {@code @XmlAttr}, {@code @XmlChild}, and
 * {@code @XmlText}. The remaining annotations from the runtime are detected and rejected
 * with a clear ERROR diagnostic.
 */
@SupportedSourceVersion(SourceVersion.RELEASE_17)
@SupportedAnnotationTypes("xmlfluss.XmlRecord")
public final class XmlDslProcessor extends AbstractProcessor {

    @Override
    public boolean process(Set<? extends TypeElement> annotations, RoundEnvironment roundEnv) {
        if (annotations.isEmpty()) return false;
        annotations
                .stream()
                .flatMap(anno -> roundEnv.getElementsAnnotatedWith(anno).stream())
                .forEach(target -> {
                            if (!(target instanceof TypeElement typeElement)) {
                                processingEnv.getMessager().printMessage(
                                        Diagnostic.Kind.ERROR,
                                        "xml-fluss-apt: @XmlRecord must annotate a record type",
                                        target);
                                return;
                            }
                            if (typeElement.getKind() != javax.lang.model.element.ElementKind.RECORD
                                    && !(typeElement.getModifiers()
                                    .contains(javax.lang.model.element.Modifier.SEALED))) {
                                processingEnv.getMessager().printMessage(
                                        Diagnostic.Kind.ERROR,
                                        "@XmlRecord requires a record type, got " + typeElement.getKind(),
                                        typeElement);
                                return;
                            }
                            RecordSpec spec;
                            NestedRegistry registry;
                            DispatchPlan plan;
                            try {
                                // Walk the record's components via the shared codegen-core classifier and consume the
                                // neutral RecordSpec / NestedRegistry directly — the emitter is now language-neutral
                                // at the model layer and renders JavaPoet types on demand via TypeRefs.
                                AptSymbolProvider sp = new AptSymbolProvider(processingEnv);
                                RecordSymbol symbol = sp.lookupRecord(typeElement.getQualifiedName().toString());
                                if (symbol == null) {
                                    // Diagnostic already emitted (e.g. @XmlNs conflict in AptSymbolProvider).
                                    return;
                                }
                                CoreClassifier core = new CoreClassifier(sp);
                                spec = core.classify(symbol);
                                if (spec == null) {
                                    // Validation diagnostics already emitted via sp.diagnostics(); skip this record.
                                    return;
                                }
                                registry = core.registry();
                                plan = DispatchPlanBuilder.build(spec, registry);
                            } catch (RuntimeException unexpected) {
                                processingEnv.getMessager().printMessage(
                                        Diagnostic.Kind.ERROR,
                                        "xml-fluss-apt: internal error processing " + typeElement + ": " + unexpected,
                                        typeElement);
                                return;
                            }
                            try {
                                // Code-generate `<Cls>Parser.java` from the RecordSpec — emits the Stream<T> parse(InputStream) entry point
                                // plus per-field state machines for attr/child/text/map handling, and writes through the Filer.
                                new Emitter(processingEnv).emit(spec, plan, registry);
                            } catch (RuntimeException unexpected) {
                                processingEnv.getMessager().printMessage(
                                        Diagnostic.Kind.ERROR,
                                        "xml-fluss-apt: failed to emit parser for " + typeElement + ": " + unexpected,
                                        typeElement);
                            }
                        }

                );
        return true;
    }
}

File: xml-fluss-ksp/src/main/kotlin/xmlfluss/ksp/XmlDslProcessor.kt
package xmlfluss.ksp

import com.google.devtools.ksp.processing.CodeGenerator
import com.google.devtools.ksp.processing.KSPLogger
import com.google.devtools.ksp.processing.Resolver
import com.google.devtools.ksp.processing.SymbolProcessor
import com.google.devtools.ksp.processing.SymbolProcessorEnvironment
import com.google.devtools.ksp.symbol.KSAnnotated
import com.google.devtools.ksp.symbol.KSClassDeclaration
import com.google.devtools.ksp.symbol.Modifier
import xmlfluss.codegen.classify.CoreClassifier
import xmlfluss.codegen.plan.DispatchPlanBuilder

/**
 * KSP processor that turns `@XmlRecord` data classes into streaming parsers.
 *
 * For each annotated class the processor:
 *
 * 1. Resolves `@XmlNs` declarations and compiles the record path with
 *    [xmlfluss.runtime.Paths.compile].
 * 2. Walks the primary constructor parameters, classifying each as `@XmlAttr`, `@XmlChild`,
 *    `@XmlText`, `@XmlMap`, or polymorphic (sealed parent annotated with `@XmlPolymorphic` +
 *    `@XmlSubtype` data-class variants), and resolves the field's coercion (scalar, temporal,
 *    decimal, custom converter, nested data class, map aggregate, or sealed dispatch).
 * 3. Emits a `${ClassName}Parser` Kotlin object via KotlinPoet, exposing
 *    `parse(InputStream): Flow<T>` plus private helpers for nested types.
 *
 * Generated parsers drive [xmlfluss.runtime.XmlReadCursor]. No reflection is used at runtime.
 *
 * The class itself is intentionally thin: the heavy lifting lives in the per-concern emitter
 * files in this package (see `EmitFile.kt`, `EmitInstance.kt`, `ChildSwitch.kt`, etc.).
 */
class XmlDslProcessor(env: SymbolProcessorEnvironment) : SymbolProcessor {

    private val codeGen: CodeGenerator = env.codeGenerator
    private val logger: KSPLogger = env.logger

    override fun process(resolver: Resolver): List<KSAnnotated> {
        val symbols = resolver.getSymbolsWithAnnotation(XML_RECORD_FQ)
            .filterIsInstance<KSClassDeclaration>()
            .toList()

        val symbolProvider = xmlfluss.ksp.spi.KspSymbolProvider(resolver, logger)
        for (cls in symbols) {
            try {
                // Drive the field model through the shared codegen-core classifier and walk
                // the neutral RecordSpec directly — TypeRefs handles JavaPoet/KotlinPoet
                // conversions on demand at emit sites.
                generate(cls, symbolProvider)
            } catch (e: ProcessorValidationException) {
                logger.error("xml-fluss-ksp: ${e.message}", cls)
            } catch (e: Throwable) {
                logger.error("xml-fluss-ksp: ${e.stackTraceToString()}", cls)
            }
        }
        return emptyList()
    }

    private fun generate(cls: KSClassDeclaration, symbolProvider: xmlfluss.ksp.spi.KspSymbolProvider) {
        // The SPI's KspSymbolProvider.lookupRecord() owns the kind check (data class OR sealed
        // parent for @XmlPolymorphic); a non-matching kind returns null after emitting the
        // diagnostic. Widen the local guard accordingly so sealed-root @XmlRecord cases don't
        // get rejected here before reaching CoreClassifier.
        if (Modifier.DATA !in cls.modifiers && Modifier.SEALED !in cls.modifiers) {
            logger.error("@XmlRecord requires data class", cls)
            return
        }
        val fqn = cls.qualifiedName?.asString() ?: return
        val symbol = symbolProvider.lookupRecord(fqn) ?: return
        val core = CoreClassifier(symbolProvider)
        val coreSpec = core.classify(symbol) ?: return
        val plan = DispatchPlanBuilder.build(coreSpec, core.registry())
        emitFile(coreSpec, plan, core.registry(), codeGen)
    }
}
